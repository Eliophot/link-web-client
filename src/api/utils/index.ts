import { AxiosError } from 'axios';
import {
  ErrorTypeV2,
  ErrorTypeV3,
  InvalidArgumentError,
  InvalidShortUrlDeletion,
  ProblemDetailsError,
  RegularNotFound,
} from '../types/errors';

const isProblemDetails = (e: unknown): e is ProblemDetailsError =>
  !!e && typeof e === 'object' && Object.keys(e).every((key) => ['type', 'detail', 'title', 'status'].includes(key));

const isAxiosError = (e: unknown): e is AxiosError<ProblemDetailsError> => !!e && typeof e === 'object' && 'response' in e;

export const parseApiError = (e: unknown): ProblemDetailsError | undefined => {
  if (isProblemDetails(e)) {
    return e;
  }

  return (isAxiosError(e) ? e.response?.data : undefined);
};

export const isInvalidArgumentError = (error?: ProblemDetailsError): error is InvalidArgumentError =>
  error?.type === ErrorTypeV2.INVALID_ARGUMENT || error?.type === ErrorTypeV3.INVALID_ARGUMENT;

export const isInvalidDeletionError = (error?: ProblemDetailsError): error is InvalidShortUrlDeletion =>
  error?.type === 'INVALID_SHORTCODE_DELETION'
  || error?.type === ErrorTypeV2.INVALID_SHORT_URL_DELETION
  || error?.type === ErrorTypeV3.INVALID_SHORT_URL_DELETION;

export const isRegularNotFound = (error?: ProblemDetailsError): error is RegularNotFound =>
  (error?.type === ErrorTypeV2.NOT_FOUND || error?.type === ErrorTypeV3.NOT_FOUND) && error?.status === 404;
