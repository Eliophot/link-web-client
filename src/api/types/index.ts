import { Visit } from '../../visits/types';
import { OptionalString } from '../../utils/utils';
import { ShortUrl, ShortUrlMeta } from '../../short-urls/data';
import { Order } from '../../utils/helpers/ordering';

export interface ShlinkShortUrlsResponse {
  data: ShortUrl[];
  pagination: ShlinkPaginator;
}

export interface ShlinkMercureInfo {
  token: string;
  mercureHubUrl: string;
}

export interface ShlinkHealth {
  status: 'pass' | 'fail';
  version: string;
}

interface ShlinkTagsStats {
  tag: string;
  shortUrlsCount: number;
  visitsCount: number;
}

export interface ShlinkTags {
  tags: string[];
  stats: ShlinkTagsStats[];
}

export interface ShlinkTagsResponse {
  data: string[];
  stats: ShlinkTagsStats[];
}

export interface ShlinkPaginator {
  currentPage: number;
  pagesCount: number;
  totalItems: number;
}

export interface ShlinkVisits {
  data: Visit[];
  pagination: ShlinkPaginator;
}

export interface ShlinkVisitsOverview {
  visitsCount: number;
  orphanVisitsCount: number;
}

export interface ShlinkVisitsParams {
  domain?: OptionalString;
  page?: number;
  itemsPerPage?: number;
  startDate?: string;
  endDate?: string;
  excludeBots?: boolean;
}

export interface ShlinkShortUrlData extends ShortUrlMeta {
  longUrl?: string;
  title?: string;
  validateUrl?: boolean;
  tags?: string[];
}

export interface ShlinkDomainRedirects {
  baseUrlRedirect: string | null;
  regular404Redirect: string | null;
  invalidShortUrlRedirect: string | null;
}

export interface ShlinkEditDomainRedirects extends Partial<ShlinkDomainRedirects> {
  domain: string;
}

export interface ShlinkDomain {
  domain: string;
  isDefault: boolean;
  redirects?: ShlinkDomainRedirects; // Optional only for Shlink older than 2.8
}

export interface ShlinkDomainsResponse {
  data: ShlinkDomain[];
  defaultRedirects?: ShlinkDomainRedirects; // Optional only for Shlink older than 2.10
}

export type TagsFilteringMode = 'all' | 'any';

type ShlinkShortUrlsOrderableFields = 'dateCreated' | 'shortCode' | 'longUrl' | 'title' | 'visits' | 'nonBotVisits';

export type ShlinkShortUrlsOrder = Order<ShlinkShortUrlsOrderableFields>;

export interface ShlinkShortUrlsListParams {
  page?: string;
  itemsPerPage?: number;
  tags?: string[];
  searchTerm?: string;
  startDate?: string;
  endDate?: string;
  orderBy?: ShlinkShortUrlsOrder;
  tagsMode?: TagsFilteringMode;
}

export interface ShlinkShortUrlsListNormalizedParams extends Omit<ShlinkShortUrlsListParams, 'orderBy'> {
  orderBy?: string;
}
