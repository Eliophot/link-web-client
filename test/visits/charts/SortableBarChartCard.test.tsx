import { shallow, ShallowWrapper } from 'enzyme';
import { range } from 'ramda';
import SortingDropdown from '../../../src/utils/SortingDropdown';
import PaginationDropdown from '../../../src/utils/PaginationDropdown';
import { OrderDir, rangeOf } from '../../../src/utils/utils';
import { Stats } from '../../../src/visits/types';
import { SortableBarChartCard } from '../../../src/visits/charts/SortableBarChartCard';
import { HorizontalBarChart } from '../../../src/visits/charts/HorizontalBarChart';

describe('<SortableBarChartCard />', () => {
  let wrapper: ShallowWrapper;
  const sortingItems = {
    name: 'Name',
    amount: 'Amount',
  };
  const stats = {
    Foo: 100,
    Bar: 50,
  };
  const createWrapper = (withPagination = false, extraStats = {}) => {
    wrapper = shallow(
      <SortableBarChartCard
        title="Foo"
        stats={{ ...stats, ...extraStats }}
        sortingItems={sortingItems}
        withPagination={withPagination}
      />,
    );

    return wrapper;
  };

  afterEach(() => wrapper?.unmount());

  it('renders stats unchanged when no ordering is set', () => {
    const wrapper = createWrapper();
    const chart = wrapper.find(HorizontalBarChart);

    expect(chart.prop('stats')).toEqual(stats);
  });

  describe('renders properly ordered stats when ordering is set', () => {
    let assert: (sortName: string, sortDir: OrderDir, keys: string[], values: number[], done: Function) => void;

    beforeEach(() => {
      const wrapper = createWrapper();
      const dropdown = wrapper.renderProp('title' as never)().find(SortingDropdown);

      assert = (sortName: string, sortDir: OrderDir, keys: string[], values: number[], done: Function) => {
        dropdown.prop('onChange')(sortName, sortDir);
        setImmediate(() => {
          const stats = wrapper.find(HorizontalBarChart).prop('stats');

          expect(Object.keys(stats)).toEqual(keys);
          expect(Object.values(stats)).toEqual(values);
          done();
        });
      };
    });

    it('name - ASC', (done) => assert('name', 'ASC', [ 'Bar', 'Foo' ], [ 50, 100 ], done));
    it('name - DESC', (done) => assert('name', 'DESC', [ 'Foo', 'Bar' ], [ 100, 50 ], done));
    it('value - ASC', (done) => assert('value', 'ASC', [ 'Bar', 'Foo' ], [ 50, 100 ], done));
    it('value - DESC', (done) => assert('value', 'DESC', [ 'Foo', 'Bar' ], [ 100, 50 ], done));
  });

  describe('renders properly paginated stats when pagination is set', () => {
    let assert: (itemsPerPage: number, expectedStats: string[], done: Function) => void;

    beforeEach(() => {
      const wrapper = createWrapper(true, range(1, 159).reduce<Stats>((accum, value) => {
        accum[`key_${value}`] = value;

        return accum;
      }, {}));
      const dropdown = wrapper.renderProp('title' as never)().find(PaginationDropdown);

      assert = (itemsPerPage: number, expectedStats: string[], done: Function) => {
        dropdown.prop('setValue')(itemsPerPage);
        setImmediate(() => {
          const stats = wrapper.find(HorizontalBarChart).prop('stats');

          expect(Object.keys(stats)).toEqual(expectedStats);
          done();
        });
      };
    });

    const buildExpected = (size: number): string[] => [ 'Foo', 'Bar', ...rangeOf(size - 2, (i) => `key_${i}`) ];

    it('50 items per page', (done) => assert(50, buildExpected(50), done));
    it('100 items per page', (done) => assert(100, buildExpected(100), done));
    it('200 items per page', (done) => assert(200, buildExpected(160), done));
    it('500 items per page', (done) => assert(500, buildExpected(160), done));
  });

  it('renders extra header content', () => {
    const wrapper = shallow(
      <span>
        <SortableBarChartCard
          title="Foo"
          stats={stats}
          sortingItems={sortingItems}
          extraHeaderContent={() => (
            <span>
              <span className="foo-span">Foo</span>
              <span className="bar-span">Bar</span>
            </span>
          )}
        />
      </span>,
    ).find(SortableBarChartCard);
    const header = wrapper.renderProp('extraHeaderContent')();

    expect(header.find('.foo-span')).toHaveLength(1);
    expect(header.find('.bar-span')).toHaveLength(1);
  });
});
