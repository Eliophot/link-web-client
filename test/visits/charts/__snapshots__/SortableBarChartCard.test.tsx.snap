// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortableBarChartCard /> renders properly ordered stats when ordering is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 288h-95.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H416c17.67 0 32-14.33 32-32S433.7 288 416 288zM352 416h-32c-17.67 0-32 14.33-32 32s14.33 32 32 32h32c17.67 0 31.1-14.33 31.1-32S369.7 416 352 416zM480 160h-159.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H480c17.67 0 32-14.33 32-32S497.7 160 480 160zM544 32h-223.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H544c17.67 0 32-14.33 32-32S561.7 32 544 32zM151.6 41.95c-12.12-13.26-35.06-13.26-47.19 0l-87.1 96.09C4.475 151.1 5.35 171.4 18.38 183.3c6.141 5.629 13.89 8.414 21.61 8.414c8.672 0 17.3-3.504 23.61-10.39L96 145.9v302C96 465.7 110.3 480 128 480s32-14.33 32-32.03V145.9L192.4 181.3C204.4 194.3 224.6 195.3 237.6 183.3c13.03-11.95 13.9-32.22 1.969-45.27L151.6 41.95z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly ordered stats when ordering is set 2`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 288h-95.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H416c17.67 0 32-14.33 32-32S433.7 288 416 288zM352 416h-32c-17.67 0-32 14.33-32 32s14.33 32 32 32h32c17.67 0 31.1-14.33 31.1-32S369.7 416 352 416zM480 160h-159.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H480c17.67 0 32-14.33 32-32S497.7 160 480 160zM544 32h-223.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H544c17.67 0 32-14.33 32-32S561.7 32 544 32zM151.6 41.95c-12.12-13.26-35.06-13.26-47.19 0l-87.1 96.09C4.475 151.1 5.35 171.4 18.38 183.3c6.141 5.629 13.89 8.414 21.61 8.414c8.672 0 17.3-3.504 23.61-10.39L96 145.9v302C96 465.7 110.3 480 128 480s32-14.33 32-32.03V145.9L192.4 181.3C204.4 194.3 224.6 195.3 237.6 183.3c13.03-11.95 13.9-32.22 1.969-45.27L151.6 41.95z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly ordered stats when ordering is set 3`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 288h-95.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H416c17.67 0 32-14.33 32-32S433.7 288 416 288zM352 416h-32c-17.67 0-32 14.33-32 32s14.33 32 32 32h32c17.67 0 31.1-14.33 31.1-32S369.7 416 352 416zM480 160h-159.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H480c17.67 0 32-14.33 32-32S497.7 160 480 160zM544 32h-223.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H544c17.67 0 32-14.33 32-32S561.7 32 544 32zM151.6 41.95c-12.12-13.26-35.06-13.26-47.19 0l-87.1 96.09C4.475 151.1 5.35 171.4 18.38 183.3c6.141 5.629 13.89 8.414 21.61 8.414c8.672 0 17.3-3.504 23.61-10.39L96 145.9v302C96 465.7 110.3 480 128 480s32-14.33 32-32.03V145.9L192.4 181.3C204.4 194.3 224.6 195.3 237.6 183.3c13.03-11.95 13.9-32.22 1.969-45.27L151.6 41.95z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly ordered stats when ordering is set 4`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 288h-95.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H416c17.67 0 32-14.33 32-32S433.7 288 416 288zM352 416h-32c-17.67 0-32 14.33-32 32s14.33 32 32 32h32c17.67 0 31.1-14.33 31.1-32S369.7 416 352 416zM480 160h-159.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H480c17.67 0 32-14.33 32-32S497.7 160 480 160zM544 32h-223.1c-17.67 0-32 14.33-32 32s14.33 32 32 32H544c17.67 0 32-14.33 32-32S561.7 32 544 32zM151.6 41.95c-12.12-13.26-35.06-13.26-47.19 0l-87.1 96.09C4.475 151.1 5.35 171.4 18.38 183.3c6.141 5.629 13.89 8.414 21.61 8.414c8.672 0 17.3-3.504 23.61-10.39L96 145.9v302C96 465.7 110.3 480 128 480s32-14.33 32-32.03V145.9L192.4 181.3C204.4 194.3 224.6 195.3 237.6 183.3c13.03-11.95 13.9-32.22 1.969-45.27L151.6 41.95z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly paginated stats when pagination is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
  <div
    class="chart-card__footer--sticky card-footer"
  >
    <nav
      aria-label="pagination"
      class=""
    >
      <ul
        class="flex-wrap mb-0 simple-paginator justify-content-center pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            aria-label="Previous"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item active"
        >
          <span
            class="page-link"
            role="link"
          >
            1
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            2
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            3
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            4
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            aria-label="Next"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly paginated stats when pagination is set 2`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
  <div
    class="chart-card__footer--sticky card-footer"
  >
    <nav
      aria-label="pagination"
      class=""
    >
      <ul
        class="flex-wrap mb-0 simple-paginator justify-content-center pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            aria-label="Previous"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item active"
        >
          <span
            class="page-link"
            role="link"
          >
            1
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            2
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            aria-label="Next"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly paginated stats when pagination is set 3`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders properly paginated stats when pagination is set 4`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> renders stats unchanged when no ordering is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <canvas
      height="0"
      role="img"
      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
      width="0"
    />
  </div>
</div>
`;
